### PR TITLE
Add features that allow the notion Client to retry api calls after be…

### PR DESCRIPTION
**Pull request by sHermanGriffiths**
…ing rate limited

# Author's Checklist

Before assigning reviewers, step through the following checklist. Check each item once it's completed. If an item is skipped, write an explanation below the item.

- [x] **Construction:** Review the code diff and confirm there's no dead code, debug code, or unnecessary comments.
- [x] **Requirements:** Review the requirements in the issue for this PR and confirm they're all implemented.
- [x] **Architecture:** Confirm that the changes to the code are consistent with the existing architecture.
- [ ] **SOUP/OTS Software:** Any new dependencies have appropriate licenses and are documented or pinned in requirements files.
  N\A
- [x] **Unit Test:** Atomic unit tests have been added, if appropriate.
- [ ] **End-to-End Tests:** End-to-end tests have been extended or updated, if appropriate.
  N\A
- [ ] **Low-level Documentation:** Comments or doc strings for affected code have been updated.
  N\A
- [ ] **High-Level Documentation:** The README is updated, as appropriate.
  N\A
- [x] **Change Log:** New features or bug fixes have been added to the change log in the README.

### CODE REVIEW

File: n2y/notion.py [hunk 1]
```diff
@@ -84,6 +116,21 @@ def __init__(
         self.load_plugins(plugins)
         self.plugin_data = {}
 
+    @property
+    def retry_api_calls(self):
+        return self.max_retries > self.retry_count
+    
+    @property
+    def max_retries(self):
```
Comment (johndgiese) [inline n2y/notion.py:124]: @bimbashrestha and @sHermanGriffiths I was thinking the "max retries" should apply to a single API call. Thus, if an API call gets a 429, it will retry that same API call up to 5 times.

If the next API call after that gets a 429, it could also get up to 5 retries before an exception is raised. Is that the behavior you guys were expecting?

Comment (sHermanGriffiths) [reply n2y/notion.py:125]: Yes, I forgot to reset the count. Just finished doing that

File: n2y/main.py [hunk 2]
```diff
@@ -24,6 +24,13 @@ def main(raw_args, access_token):
         formatter_class=argparse.RawTextHelpFormatter,
     )
     parser.add_argument("config", help="The path to the config file")
+    parser.add_argument(
+        '--max_retries', '-m', default=5,
```
Comment (bimbashrestha) [inline n2y/main.py:28]: We have the default set to 5 in multiple places. Let's extract the 5 to a constant somewhere and use that everywhere. That way, we can just change it in one place when we need to.

File: n2y/notion.py [hunk 3]
```diff
@@ -46,6 +47,35 @@
 # TODO: Rename this file `client.py`
 
 
+def retry_api_call(api_call):
```
Comment (bimbashrestha) [inline n2y/notion.py:50]: I like that you're using a decorator. Nice idea!

I know there are a few popular python libraries that provide a retry decorator ([retry](https://github.com/invl/retry), [tenacity](https://github.com/jd/tenacity) are two that I know of). I think we should use one of them instead of implementing our own. You can probably swap their decorator for yours as-is. It'll also make it easier for us to add more advanced retry functionality like backoffs, jitter, choose which exceptions to retry on, etc later on. What do you think @sHermanGriffiths and @johndgiese?

Also @sHermanGriffiths, if we do implement our own decorator, we need a good unit test to make sure it works.

File: n2y/notion.py [hunk 4]
```diff
@@ -84,6 +117,21 @@ def __init__(
...
+    def retry_api_calls(self):
+        return self.max_retries > self.retry_count
+
+    @property
+    def max_retries(self):
+        return self._max_retries
+
+    @max_retries.setter
+    def max_retries(self, val):
+        if int(val) > 5:
```
Comment (bimbashrestha) [inline n2y/notion.py:130]: Extract constant max retries.

File: n2y/notion.py [hunk 5]
```diff
@@ -66,10 +96,13 @@ def __init__(
         media_root='.',
         media_url='',
         plugins=None,
+        max_retries=5,
```
Comment (bimbashrestha) [inline n2y/notion.py:99]: Extract constant max retries

File: n2y/rich_text.py [hunk 6]
```diff
@@ -24,10 +24,9 @@ class RichText:
     a property_value or somewhere else.
     """
 
-    def __init__(self, client, notion_data, block=None, table=False):
+    def __init__(self, client, notion_data, block=None):
```
Comment (bimbashrestha) [inline n2y/rich_text.py:27]: Is this change related to retries?

File: tests/test_plain_text_to_pandoc.py [hunk 7]
```diff
@@ -21,8 +21,3 @@ def test_plain_text_to_pandoc_spaces_after():
 def test_plain_text_to_pandoc_spaces_after_newline():
     pandoc_ast = RichText.plain_text_to_pandoc("hello\n  world")
     assert pandoc_ast == [Str("hello"), LineBreak(), Space(), Space(), Str("world")]
-
-
-def test_plain_text_to_pandoc_table():
```
Comment (bimbashrestha) [inline tests/test_plain_text_to_pandoc.py:26]: Why are we deleting this test case?

**Review by bimbashrestha — CHANGES_REQUESTED** [review]
Nice start @sHermanGriffiths! Requested some changes.

Also, I noticed that this PR contains changes for more than just retries. Let's move those to their own PR off of `main` so that it can be reviewed on it's own. 

Combining multiple things into a PR makes sense if they are all small items or if they are related to each other. The retry stuff seems pretty independent from the table stuff. If they're not, let me know.

File: n2y/notion.py (hunk 1 above)
Comment (bimbashrestha) [reply n2y/notion.py:125]: Nice catch. Yeah, each api call should definitely be treated independently in terms of retries.

File: n2y/notion.py (hunk 3 above)
Comment (johndgiese) [reply n2y/notion.py:50]: I like the idea of using a library for this.

Comment (sHermanGriffiths) [reply n2y/notion.py:50]: I like the idea of using a library as well, but I can't find any that parse the delay time from the exception itself, which is the premise of the functionality we're trying to create here.

Comment (johndgiese) [reply n2y/notion.py:50]: @sHermanGriffiths that's a great point, I think we should use the header sent back from Notion to figure out the timeout time.

Comment (bimbashrestha) [reply n2y/notion.py:50]: @sHermanGriffiths Ah I see. Great point and that makes total sense!

I'll leave the decision up to you. If you do move forward with a custom implementation though, please include a test for it.

File: tests/test_utils.py [hunk 8]
```diff
@@ -43,3 +45,69 @@ def test_page_id_from_share_link():
...
+        elif client.retry_count == 3:
+            assert isclose(14, seconds, abs_tol=1)
+            raise APIResponseError(Resp(2), '', 439)
+    client = Client('', max_retries=3)
+    try:
+        tester(client, datetime.now())
+    except APIResponseError:
+        assert True
+    else:
+        assert False
```
Comment (bimbashrestha) [inline tests/test_utils.py:113]:
```suggestion
    with pytest.raises(APIResponseError):
        tester(client, datetime.now())
```

File: n2y/utils.py [hunk 9]
```diff
@@ -34,10 +37,6 @@ def processed_date_to_plain_text(processed_date):
         return processed_date
 
 
-# see https://pandoc.org/MANUAL.html#exit-codes
-PANDOC_PARSE_ERROR = 64
```
Comment (bimbashrestha) [inline n2y/utils.py:38]: Is deleting this intentional?

**Review by bimbashrestha — CHANGES_REQUESTED** [review]
@sHermanGriffiths Nice work! 

This is close. I left some inline comments (mostly small). 

Let's also add a test for making sure retry counts are independent for each api call. The thing David mentioned. You can just call trigger your decorator two times in a row and make sure the retry count is reset for the second run.

File: tests/test_utils.py [hunk 10]
```diff
@@ -43,3 +45,69 @@ def test_page_id_from_share_link():
...
     assert id_from_share_link(share_link) == page_id
+
+
+def test_retry_api_call_no_error():
+    @retry_api_call
+    def tester(client):
+        assert client.max_retries == 5
+        assert client.retry_count == 0
+        assert client.retry_api_calls
+        return 5
```
[the remark this answers is not in the document]
Comment (sHermanGriffiths) [reply tests/test_utils.py:56]: Yes, I've changed it to just return true

File: n2y/utils.py (hunk 9 above)
Comment (sHermanGriffiths) [reply n2y/utils.py:38]: not at all

**Review by bimbashrestha — APPROVED** [review]
Nice work!

**Review by bimbashrestha — CHANGES_REQUESTED** [review]
Ah actually looks like there are some CI failures, so I'm re-requesting changes

**Review by bimbashrestha — APPROVED** [review]
